### ISSUE TITLE
Linking issues on Windows

### ISSUE THREAD

**Issue by chaodhib** [mfontanini/cppkafka#103]
Hello,

I'm having issues to use the library on Windows  with Visual Studio and CMake. To be more specific:
- librdkafka.redist 0.11.4
- CMake 3.12.0
- Visual Studio 2017 (15.5.2)

First of all, I used [this page in the wiki](https://github.com/mfontanini/cppkafka/wiki/Compiling) to compile the library statically, into a lib file. It looks like there is one error in the page though: For RDKAFKA_LIBRARY, i had to use `NuGet/librdkafka.redist.0.11.4/build/native/lib/win7/x64/win7-x64-Release/v120/librdkafkacpp.lib` (and not the `runtimes` folder, since it contains the DLLs)

Once the library is compiled, I created a very basic main.cpp file:

````C++
#include "configuration.h"

int main(int argc, char* argv[])
{
    // Construct the configuration
    cppkafka::Configuration config = {
        { "metadata.broker.list", "" },
    { "group.id", 1 },
    // Disable auto commit
    { "enable.auto.commit", false }
    };
}
````

Then I tried to run it. Unfortunately, it doesn't link properly. I can clearly see the library file mentioned during the link process:
`C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.12.25827\bin\HostX86\x64\link.exe /ERRORREPORT:PROMPT /OUT:"C:\Users\chaouki\Desktop\Bachelier Info\BA3\Epreuve integrée\kafka-integration\build\Debug\kafka-integration.exe" /INCREMENTAL /NOLOGO C:\Users\chaouki\Documents\Git\cppkafka\build\lib\Release\cppkafka.lib "C:\local\boost_1_66_0\lib64-msvc-14.0\libboost_filesystem-vc141-mt-gd-x64-1_66.lib" "C:\local\boost_1_66_0\lib64-msvc-14.0\libboost_program_options-vc141-mt-gd-x64-1_66.lib" "C:\local\boost_1_66_0\lib64-msvc-14.0\libboost_system-vc141-mt-gd-x64-1_66.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:/Users/chaouki/Desktop/Bachelier Info/BA3/Epreuve integrée/kafka-integration/build/Debug/kafka-integration.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/Users/chaouki/Desktop/Bachelier Info/BA3/Epreuve integrée/kafka-integration/build/Debug/kafka-integration.lib" /MACHINE:X64  /machine:x64 "kafka-integration.dir\Debug\main.obj"`

So I assume that my CMake config is not at fault here. Any idea what could be the issue?

Thanks in advance for the help

**Comment by mfontanini** [comment]
Can you post here the linker errors?

**Comment by chaodhib** [comment]
Sure. Here is the full output: [HERE](https://gist.github.com/chaodhib/61567a7edd460f6a61cce9ce72685543#file-gistfile1-txt)

**Comment by mfontanini** [comment]
I'll add this to that wiki entry but you should define `CPPKAFKA_STATIC` as a preprocessor macro when importing cppkafka into your project when you build it as a static lib. This is a common pattern used in many libraries. This is only an issue on Windows caused by the way you need to explicitly every exported symbol in a library.

**Comment by chaodhib** [comment]
Could you be more specific? You mean adding `#define CPPKAFKA_STATIC 1` after the includes. And `#undef CPPKAFKA_STATIC` at the end? Just tried it. Doesn't seem to work.

**Comment by mfontanini** [comment]
I mean add that in visual studio as a preprocessor macro (down somewhere in project -> properties, c++... I don't remember now). e.g. `CPPKAFKA_STATIC=1`

**Comment by chaodhib** [comment]
With 
![image](https://user-images.githubusercontent.com/6612710/42914642-ac952a1e-8afb-11e8-97ff-bec9a1ccd6b3.png)

I get the same link errors: [full output](https://gist.github.com/chaodhib/df197158f68422a45f2c1a376b2a27c9#file-gistfile1-txt)

**Comment by mfontanini** [comment]
I imagine you're linking your application with rdkafka.lib as well, right?

**Comment by chaodhib** [comment]
My understanding was that the content of rdkafka.lib will be included in cppkafka.lib. Was I wrong?

**Comment by chaodhib** [comment]
Anyway, I included rdkafka.lib using CMake. Sadly, nothing changes: [output ](https://gist.github.com/chaodhib/2c289b1af3f24b8edb867b24e0e47a12#file-gistfile1-txt) (i increased the verbosity)

**Comment by mfontanini** [comment]
You are including `C:\NuGet\librdkafka.redist.0.11.4\build\native\lib\win7\x64\win7-x64-Release\v120\librdkafkacpp.lib` and that's not what you want. You want rdkafka.lib, not rdkafkacpp.lib.

**Comment by chaodhib** [comment]
There are still errors but significantly less: [output](https://gist.github.com/chaodhib/992162d125ce6924acb9e68afb41feda#file-gistfile1-txt)

**Comment by jlohiya** [comment]
Hi @chaodhib

Did it work for you? Could you please share your code if it worked?

**Comment by chaodhib** [comment]
Hi @jlohiya 

Sadly no. It hasn't work yet. On the last attempt, there was still 2 linkage errors (cf my previous comment for the full log output).

**Comment by chaodhib** [comment]
Yeah sorry. Actually, I didn't had the chance to work on this since yesterday.

Anyway, what you said does indeed get rid of the remaining linkage errors. It now correctly produces an executable file. However, when launched, the file throw an error similar to what is [here](https://stackoverflow.com/questions/38592708/why-does-my-program-ask-for-opencvs-dll-even-though-i-link-statically-against): Basically, "librdkafka.dll is missing from your computer". I checked the librdkafka.lib that I use. And indeed, it's actually an import lib. Copying the DLL of librdkafka in the same folder as the executable did the trick.

Thanks for the help!
